### PR TITLE
rose docs: broken link detection

### PR DESCRIPTION
**Pull request by oliver-sanders**
Added a test to detect broken links. Detects:
- External Links
    - Page exists and returns a sensible HTTP code.
- Internal Links
    - Page exists
    - URL fragments (e.g. example.com#fragment) reference sections that are present

Fixed many broken links.

Caveats:
- At present some URLs fail for SSL reasons, the crude solution implemented is to retry the affected URLs with SSL verification turned **OFF**! As this testing will only be done on URLs we have put into the docs. If you think this is a problem I can spend some time trying to get around it.
- Does not check URL fragments on external sites (would require downloading the whole page).

@matthewrmshin @arjclark Please Review.

### CODE REVIEW

File: t/docs/lib/python/urlvalidator.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
+"""A crude URL validator"""
```
Comment (matthewrmshin) [inline t/docs/lib/python/urlvalidator.py]: Please add usual header.

File: t/docs/lib/python/urlvalidator.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    report_invalid_urls(int_errors, ext_errors, frg_errors, malformed)
+
+    if int_errors or ext_errors or frg_errors or malformed:
+        sys.exit(1)
+    sys.exit(0)
+
+if __name__ == '__main__':
+    if len(sys.argv) != 2:
+        print ('Usage: Please supply root directory to validate the contents '
+               'of.')
```
Comment (matthewrmshin) [inline t/docs/lib/python/urlvalidator.py]: Don't think you need a usage check, given that it is only used in our tests!

File: t/docs/lib/python/urlvalidator.py (hunk 1 above)
Comment (oliver-sanders) [inline t/docs/lib/python/urlvalidator.py]: Done.

File: t/docs/lib/python/urlvalidator.py (hunk 2 above)
Comment (oliver-sanders) [inline t/docs/lib/python/urlvalidator.py]: Might as well check we have the correct number of arguments and provide some form of CLI documentation to future developers.

Comment (matthewrmshin) [inline t/docs/lib/python/urlvalidator.py]: In which case, it ought to go to STDERR, or use `sys.exit(...)`.

**Comment by oliver-sanders** [comment]
Added a test for the rose macro (advanced) tutorial.

File: t/docs/02-macro-tutorial.t [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#
+# Rose is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Rose. If not, see <http://www.gnu.org/licenses/>.
+#-------------------------------------------------------------------------------
+# Test "rose date".
```
Comment (arjclark) [inline t/docs/02-macro-tutorial.t]: Needs changing to what the test is.

File: t/docs/01-validate-validator.t [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+#
+# Rose is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Rose. If not, see <http://www.gnu.org/licenses/>.
+#-------------------------------------------------------------------------------
+# Test "rose date".
```
Comment (arjclark) [inline t/docs/01-validate-validator.t]: Needs changing to what the test is

File: t/docs/00-urls.t [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+#
+# Rose is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Rose. If not, see <http://www.gnu.org/licenses/>.
+#-------------------------------------------------------------------------------
+# Test "rose date".
```
Comment (arjclark) [inline t/docs/00-urls.t]: Needs changing to what the test is

**Pull request merged**